src/handshake/server.rs

```diff
@@ -83,15 +87,25 @@ impl<'a, T: AsyncRead + AsyncWrite + Unpin> Server<'a, T> {
     }
 
     /// Await an incoming client handshake request.
-    pub async fn receive_request(&mut self) -> Result<ClientRequest<'a>, Error> {
+    pub async fn receive_request<'r>(&'r mut self) -> Result<ClientRequest<'r>, Error> {
         self.buffer.clear();
+
+        let mut skip = 0;
+
         loop {
             crate::read(&mut self.socket, &mut self.buffer, BLOCK_SIZE).await?;
-            if let Parsing::Done { value, offset } = self.decode_request()? {
-                self.buffer.advance(offset);
-                return Ok(value)
+
+            // We don't expect body, so can search for the CRLF headers tail from
+            // the end of the buffer.
+            if self.buffer[skip..].windows(4).rev().find(|w| w == b"\r\n\r\n").is_some() {
```

Comment:
devil's advocate: what if someone posts a malicious handshake with 5Mb of garbage bytes as the body, won't we spend a lot of time here?

Reply:
A malicious actor could just make a connection and send 5MB of garbage bytes without any headers at all. It might make sense to set some heuristic of how many maximum bytes we'll read from the socket till we assume that this request is garbage.

Reply:
Added an 8kb limit to headers since that [seems to be the default on HTTP servers](https://stackoverflow.com/questions/686217/maximum-on-http-header-values). We can make that configurable down the line if need be.